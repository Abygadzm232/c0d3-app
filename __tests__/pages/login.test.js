jest.mock('@apollo/react-hooks')
import React from 'react'
import { render, fireEvent, wait, act } from '@testing-library/react'
import LoginPage from '../../pages/login'
import { useMutation } from '@apollo/react-hooks'

const mockFn = jest.fn()
useMutation.mockReturnValue([mockFn, {data: {}, error: {}}])

// Mock global.window
global.window = Object.create(window)

// define property to be modifed beforeEach test
Object.defineProperty(global.window, 'location', {
  value: { pathname: '/' }
})

describe('Login Page', () => {
  const fillOutLoginForm = async getByTestId => {
    const usernameField = getByTestId('username')
    const passwordField = getByTestId('password')
    await wait(
      fireEvent.change(usernameField, {
        target: {
          value: 'user name'
        }
      }),
      fireEvent.change(passwordField, {
        target: {
          value: 'password123'
        }
      })
    )
  }
  beforeEach(() => {
    global.window.location.pathname = '/login' // reset path
  })

    test('should not submit when empty form', async () => {
    const { getByTestId } = render(<LoginPage />)
    const submitButton = getByTestId('submit')
    await wait(() => {
      act(() => {
        fireEvent.click(submitButton)
      })
    })
    expect(mockFn).not.toBeCalled()
  })

  test('Should redirect to curriculum', async () => {
    useMutation.mockReturnValue([mockFn, {data: {login: { success: true}}}])
    const { getByTestId } = render(<LoginPage />)
    const submitButton = getByTestId('submit')
    fillOutLoginForm(getByTestId)

    await wait(() => {
        fireEvent.click(submitButton)
    })
    expect(global.window.location.pathname).toEqual('/curriculum')
  })

    test('Should set alert visible on invalid credentials', async () => {
      useMutation.mockReturnValue([mockFn, {
        error: {
          graphQLErrors: [
            {
              message: 'UserInputError: User does not exist!'
            }
          ]
        }
      }
      ])

    const { getByTestId } = render(<LoginPage />)
    const submitButton = getByTestId('submit')
    fillOutLoginForm(getByTestId)

    act(
      () => {
        fireEvent.click(submitButton)
      },
      () => {
        getByText('User does not exist!')
      }
    )
  })
})
