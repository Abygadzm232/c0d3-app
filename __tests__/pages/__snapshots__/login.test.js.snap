// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Page Should set alert visible on invalid credentials 1`] = `
<div>
  <div
    class="row mt-5"
  >
    <div
      class="card shadow-sm col-sm-8 col-md-7 col-lg-6 col-xl-6 m-auto px-md-5 border-0"
    >
      <div
        class="card-body text-center pt-5 pb-5"
      >
        <h1
          class="card-title h2 fw-bold mb-5"
        >
          Login
        </h1>
        <p
          class="card-text"
        />
        <form
          action="#"
          data-testid="form"
        >
          <div
            class="form-group"
          >
            <div
              class="alert d-flex justify-content-between mt-3 alert-danger"
              role="alert"
            >
              <div
                class="d-flex gap-3"
              >
                <img
                  height="24"
                  src="/assets/curriculum/icons/exclamation.svg"
                  width="24"
                />
                User does not exist! 
              </div>
            </div>
            <div
              class="mt-3 d-grid"
            >
              <input
                class="form-control form-control-lg fw-light mb-3 "
                data-testid="username"
                name="username"
                placeholder="Username"
                value="fake username"
              />
              <input
                class="form-control form-control-lg fw-light mb-3 "
                data-testid="password"
                name="password"
                placeholder="Password"
                type="password"
                value="fake password"
              />
              <button
                class="btn btn-primary btn-lg btn mb-3"
                data-testid="submit"
                type="submit"
              >
                Login to Your Account
              </button>
            </div>
          </div>
        </form>
        <a
          class=" link"
          href="/forgotpassword"
        >
          Forgot your password?
        </a>
      </div>
    </div>
  </div>
</div>
`;
